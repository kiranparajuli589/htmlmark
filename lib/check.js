const lexer = [{
	type: "blockquote",
	raw: ">>> three\n>\n> 1\n>\n>  1\n> > 2\n> > > 3\n> > > > 4",
	tokens: [
		{
			type: "blockquote",
			raw: ">> three\n\n",
			tokens: [{
				type: "blockquote",
				raw: "> three\n",
				tokens: [{
					type: "paragraph",
					raw: "three\n",
					text: "three",
					tokens: [{
						type: "text",
						raw: "three",
						text: "three"
					}]
				}],
				text: "three\n"
			}],
			text: "> three\n"
		},
		{
			type: "paragraph",
			raw: "1",
			text: "1",
			tokens: [{
				type: "text",
				raw: "1",
				text: "1"
			}]
		},
		{
			type: "space",
			raw: "\n\n"
		},
		{
			type: "paragraph",
			raw: " 1\n",
			text: " 1",
			tokens: [{
				type: "text",
				raw: " 1",
				text: " 1"
			}]
		},
		{
			type: "blockquote",
			raw: "> 2\n> > 3\n> > > 4",
			tokens: [
				{
					type: "paragraph",
					raw: "2\n",
					text: "2",
					tokens: [{
						type: "text",
						raw: "2",
						text: "2"
					}]
				},
				{
					type: "blockquote",
					raw: "> 3\n> > 4",
					tokens: [
						{
							type: "paragraph",
							raw: "3\n",
							text: "3",
							tokens: [{
								type: "text",
								raw: "3",
								text: "3"
							}]
						},
						{
							type: "blockquote",
							raw: "> 4",
							tokens: [{
								type: "paragraph",
								raw: "4",
								text: "4",
								tokens: [{
									type: "text",
									raw: "4",
									text: "4"
								}]
							}],
							text: "4"
						}],
					text: "3\n> 4"
				}
			],
			text: "2\n> 3\n> > 4"
		}
	],
	text: ">> three\n\n1\n\n 1\n> 2\n> > 3\n> > > 4"
}]
