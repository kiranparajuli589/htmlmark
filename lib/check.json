[
	{
		"type": "quote",
		"depth": 3,
		"value": [
			" three"
		],
		"raw": [
			"> > > three"
		]
	},
	{
		"type": "quote",
		"depth": 1,
		"value": [
			" one",
			{
				"type": "quote",
				"depth": 2,
				"value": [
					" two",
					" # h-two",
					{
						"type": "quote",
						"depth": 3,
						"value": [
							" three",
							{
								"type": "quote",
								"depth": 4,
								"value": [
									" four"
								],
								"raw": [
									"> > > > four"
								],
								"relativeDepth": 1
							}
						],
						"raw": [
							"> > > three",
							"> > > > four"
						],
						"relativeDepth": 1
					}
				],
				"raw": [
					"> > two",
					"> > # h-two",
					"> > > three"
				],
				"relativeDepth": 1
			}
		],
		"raw": [
			"> one"
		]
	}
]
