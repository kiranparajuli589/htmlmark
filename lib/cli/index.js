import {MarkdownParser} from "../index.js";
import {File} from "../util/file.js";

const args = process.argv.slice(2);

const md = args[0];

const mdp = new MarkdownParser()

let parsed
if (File.exists(md)) {
	parsed = mdp.parseFile(md)
} else {
	parsed = mdp.parse(md)
}

console.log(parsed)
