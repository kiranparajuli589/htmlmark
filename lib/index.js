import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"


export class MarkdownParser {
	indent
	codeHighlightFn
	frontMatter
	noLinkRefUsage
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
		this.frontMatter = config.frontMatter || false
		this.noLinkRefUsage = config.noLinkRefUsage || false
	}
	tokenize(str) {
		if (typeof str !== "string") throw new Error("Input must be a string")
		const lexer = new Lexer(str.split("\n"), null, this.frontMatter)
		return lexer.run()
	}
	parse(str) {
		if (typeof str !== "string") throw new Error("Input must be a string")
		const lex = this.tokenize(str)
		const parser = new Parser(lex, null, this.codeHighlightFn)
		return parser.run()
	}
	getFrontMatter(str) {
		if (typeof str !== "string") throw new Error("Input must be a string")
		const lexer = new Lexer(str.split("\n"), null, true)
		return lexer.getFrontMatter()
	}
}
