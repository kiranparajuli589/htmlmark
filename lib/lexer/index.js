import { TOKENS } from "../util/tokens.js"
import { REGEX } from "../regex/index.js"
import { Commons } from "./tokenizer/commons.js"
import { CodeBlock } from "./tokenizer/codeblock.js"
import { Table } from "./tokenizer/table.js"
import { List } from "./tokenizer/list.js"
import { Quote } from "./tokenizer/quote.js"


export class Lexer {
	#cursor
	#lines
	#lexerData
	#currLine
	#currLineIndent
	#lastLexerItem
	#lexerLengthBefore

	#lazyTokens = [
		TOKENS.QUOTE,
		TOKENS.LIST
	]

	constructor(lines) {
		this.#lines = lines
		this.#lexerData = []
	}

	#runCurrLineLexer() {
		if (["", "\n"].includes(this.#currLine.trim())) return this.#runNewLineLexer()

		Commons.resetRegexes()

		if (REGEX.HR_LINE.test(this.#currLine)) return this.#runHrLineLexer()
		if (REGEX.COMMENT.test(this.#currLine)) return this.#runCommentLexer()
		if (REGEX.IMAGE.test(this.#currLine)) return this.#runImageLexer()
		if (REGEX.HEADING.test(this.#currLine)) return this.#runHeadingLexer()
		if (Quote.test(this.#currLine)) return this.#runQuoteLexer()
		if (List.test(this.#currLine)) return this.#runListLexer()

		if (CodeBlock.typeTest(this.#lines, this.#cursor, this.#currLineIndent)) return this.#runCodeBlockLexer()
		if (Table.typeTest(this.#lines, this.#cursor, this.#currLineIndent)) return this.#runTableLexer()

		const found = this.#checkForLazyLexer()
		if (found) return

		// otherwise, it is a paragraph
		return this.#runParagraphLexer()
	}

	#runNewLineLexer() {
		// cannot be added at the top of the content
		if (this.#lexerLengthBefore === 0) return true
		// if there are multiple new lines in a row,
		// only single new line is added to the lexerContent
		if (this.#lastLexerItem.type !== TOKENS.NEW_LINE) {
			this.#lexerData.push({
				type: TOKENS.NEW_LINE
			})
		}
	}

	#runHrLineLexer() {
		if (this.#lastLexerItem && this.#lastLexerItem.type === TOKENS.HR_LINE) return true
		this.#lexerData.push({
			type: TOKENS.HR_LINE
		})
	}

	#runCodeBlockLexer() {
		const cbTokens = CodeBlock.tokenize(this.#lines, this.#cursor, this.#currLineIndent)

		this.#lexerData.push({
			type: TOKENS.CODE_BLOCK,
			indent: this.#currLineIndent,
			language: cbTokens.language || null,
			value: cbTokens.body
		})
		// skip the lines that were parsed
		this.#cursor = cbTokens.cursor
	}

	#runTableLexer() {
		const tableTokens = Table.tokenize(this.#lines, this.#cursor, this.#currLineIndent)
		this.#lexerData.push(tableTokens.lexer)
		this.#cursor = tableTokens.cursor
	}

	#runListLexer() {
		const listTokens = List.tokenize(this.#lines, this.#cursor, this.#currLineIndent)
		this.#cursor = listTokens.cursor

		this.#lexerData.push({
			type: TOKENS.LIST,
			indent: this.#currLineIndent,
			meta: listTokens.type,
			items: listTokens.body
		})
	}

	#runQuoteLexer() {
		const quoteTokens = Quote.tokenize(this.#lines, this.#cursor, this.#currLineIndent)
		this.#cursor = quoteTokens.cursor

		this.#lexerData.push({
			type: TOKENS.QUOTE,
			indent: this.#currLineIndent,
			depth: quoteTokens.depth,
			tokens: quoteTokens.tokens,
			raw: quoteTokens.raw
		})
	}

	#runHeadingLexer() {
		const headingTokens = Commons.tokenizeHeading(this.#currLine)
		const headingLexer = {
			type: TOKENS.HEADING,
			...headingTokens,
			indent: this.#currLineIndent,
			raw: this.#currLine
		}
		this.#lexerData.push(headingLexer)
	}

	#runCommentLexer() {
		const commentTokens = Commons.tokenizeComment(this.#currLine)
		const commentLexer = {
			type: TOKENS.COMMENT,
			indent: this.#currLineIndent,
			...commentTokens,
			raw: this.#currLine
		}
		this.#lexerData.push(commentLexer)
	}

	#runImageLexer() {
		const iTokens = Commons.tokenizeImage(this.#currLine)
		const imageLexer = {
			type: TOKENS.IMAGE,
			indent: this.#currLineIndent,
			...iTokens,
			raw: this.#currLine
		}
		this.#lexerData.push(imageLexer)
	}

	#runParagraphLexer() {
		if (
			this.#lastLexerItem &&
      this.#lastLexerItem.type === TOKENS.PARAGRAPH &&
      this.#lastLexerItem.indent === this.#currLineIndent
		) {
			this.#lastLexerItem.value += ` ${this.#currLine.trim()}`
			this.#lastLexerItem.raw += `\n${this.#currLine}`
		} else {
			this.#lexerData.push({
				type: TOKENS.PARAGRAPH,
				indent: this.#currLineIndent,
				value: this.#currLine.trimEnd(),
				raw: this.#currLine
			})
		}
	}

	#checkForLazyLexer() {
		if (this.#lastLexerItem) {
			if (this.#lazyTokens.includes(this.#lastLexerItem.type)) {
				if (this.#lastLexerItem.type === TOKENS.LIST) {
					const lastItemToken = this.#lastLexerItem.items.at(-1).tokens.value.at(0)
					if (lastItemToken.type === TOKENS.PARAGRAPH) {
						lastItemToken.value += ` ${this.#currLine.trim()}`
						lastItemToken.raw += `\n${this.#currLine}`
						return true
					}
				} else if (this.#lastLexerItem.type === TOKENS.QUOTE) {
					const lastToken = this.#lastLexerItem.tokens.at(-1)
					const lastPiece = Array.isArray(lastToken) ? lastToken.at(-1) : lastToken.value.at(-1)
					if (lastPiece.type === TOKENS.PARAGRAPH) {
						lastPiece.value += ` ${this.#currLine.trim()}`
						lastPiece.raw += `\n${this.#currLine}`
						return true
					}
				}
			}
		}
		return false
	}

	run() {
		for (this.#cursor = 0; this.#cursor < this.#lines.length; this.#cursor++) {
			this.#currLine = this.#lines[this.#cursor].trimEnd()
			this.#lexerLengthBefore = this.#lexerData.length
			this.#lastLexerItem = this.#lexerData[this.#lexerLengthBefore - 1] || null
			this.#currLineIndent = Commons.getIndent(this.#currLine)

			this.#runCurrLineLexer()
		}
		return this.#lexerData
	}
}
