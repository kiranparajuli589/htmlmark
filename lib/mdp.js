import { Lexer } from "./lexer/index.js"
import { Parser } from "./parser/index.js"

/**
 * Markdown Parser for Javascript Application
 */
export class MDP {
	indent
	codeHighlightFn
	constructor(config = {}) {
		this.indent = config.indent || 4
		this.codeHighlightFn = config.codeHighlightFn || null
	}
	/**
	 * Runs the HTML parsing for the given lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	h(lines) {
		const lexer = new Lexer(lines)
		const parser = new Parser(lexer.run(), null, this.codeHighlightFn)
		return parser.run()
	}

	/**
	 * Runs HTML parsing for the lines of text
	 *
	 * @param {Array} lines - An array of lines of text
	 *
	 * @returns {string} - The parsed HTML
	 */
	hT(lines) {
		const start = Date.now()
		const html = this.h(lines)
		const end = Date.now()
		console.info(`\x1b[35mParsing time: ${end - start}ms\x1b[0m`)
		return html
	}

	/**
	 * For demo preview
	 *
	 * @param {String} str - The string to parse
	 * @param {Function|null} codeHighlightFn - A function to highlight code blocks
	 *
	 * @returns {{ elapsedTime: number, html: string, lex: Object}}
	 */
	hP(str, codeHighlightFn = null) {
		const lines = str.split("\n")
		const start = Date.now()
		const lexer = new Lexer(lines)
		const lex = lexer.run()
		const parser = new Parser(lex, null, codeHighlightFn)
		const html = parser.run()
		const end = Date.now()
		return {
			elapsedTime: end - start,
			html,
			lex
		}
	}
}
