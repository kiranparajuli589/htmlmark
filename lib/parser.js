const TOKENS = require("./tokens")

const parseParaToken = (token) => {
  let paraHtml = "%s"
  for (let j=0; j<token.tokens.length; j++) {
    const pToken = token.tokens[j]
    // now pToken can be of type
    // code, bold, italic, strikeThrough, link, text
    if (pToken.type === TOKENS.CODE) {
      paraHtml = paraHtml.replace("%s", `<code>${pToken.value}</code>%s`)
    } else if (pToken.type === TOKENS.BOLD) {
      paraHtml = paraHtml.replace("%s", `<strong>${pToken.value}</strong>%s`)
    } else if (pToken.type === TOKENS.ITALIC) {
      paraHtml = paraHtml.replace("%s", `<em>${pToken.value}</em>%s`)
    } else if (pToken.type === TOKENS.STRIKE_THROUGH) {
      paraHtml = paraHtml.replace("%s", `<s>${pToken.value}</s>%s`)
    } else if (pToken.type === TOKENS.LINK) {
      paraHtml = paraHtml.replace("%s", `<a href="${pToken.href}">${pToken.value}</a>%s`)
    } else if (pToken.type === TOKENS.TEXT) {
      paraHtml = paraHtml.replace("%s", `${pToken.value}%s`)
    }
  }
  return paraHtml.replace("%s", "")
}

const parser = (tokens) => {
    
  const parsedContent = []
  for (let i=0; i<tokens.length; i++) {
    const token = tokens[i]
    const lastItem = (parsedContent.length > 0) ? parsedContent[parsedContent.length - 1] : null
    
    // new line
    if (token.type === TOKENS.NEW_LINE) {
      parsedContent.push("<br>")
      continue
    }
    
    // codeblock
    if (token.type === TOKENS.CODE_BLOCK) {
      let cbHtml = "<pre><code%class>%s</code></pre>"
      if (token.language) {
        cbHtml = cbHtml.replace("%class", ` class="language-${token.language}"`)
      } else {
        cbHtml = cbHtml.replace("%class", "")
      }
      parsedContent.push(cbHtml.replace("%s", token.tokens[0].value))
      continue
    }
    
    // quote
    if (token.type === TOKENS.QUOTE) {
      let quoteHtml = "<blockquote>%s</blockquote>"
      for (let j=0; j<token.depth; j++) {
        quoteHtml = quoteHtml.replace("%s", "<blockquote>%s</blockquote>")
      }
      parsedContent.push(quoteHtml.replace("%s", parseParaToken(token)))
      continue
    }
    
    // heading
    if (token.type === TOKENS.HEADING) {
      parsedContent.push(`<h${token.level}>${parseParaToken(token)}</h${token.level}>`)
      continue
    }
    
    // list item
    // if the list item is not the first one
    // we need to insert it into the list created beforehand
    if ([TOKENS.LIST_ITEM, TOKENS.COUNT_ITEM].includes(token.type)) {
      const listItem = `<li>${parseParaToken(token)}</li>%s`
      if (lastItem && lastItem.trim().startsWith("<ul>")) {
        parsedContent[parsedContent.length - 1] = lastItem.replace("%s", listItem)
      } else {
        parsedContent.push(`<ul>${listItem}%s</ul>`)
      }
      continue
    }
    
    // TOKENS.checkBox, TOKENS.checkBoxChecked
    if ([TOKENS.CHECKBOX, TOKENS.CHECKBOX_CHECKED].includes(token.type)) {
      const checkboxItem = `<li><input type="checkbox" ${token.type === TOKENS.CHECKBOX_CHECKED ? "checked" : ""}>${parseParaToken(token)}</li>%s`
      if (lastItem && lastItem.trim().startsWith("<ul>")) {
        parsedContent[parsedContent.length - 1] = lastItem.replace("%s", checkboxItem)
      } else {
        parsedContent.push(`<ul>${checkboxItem}</ul>`)
      }
        
      continue
    }
    
    // comment
    if (token.type === TOKENS.COMMENT) {
      continue
    }
    
    // image
    if (token.type === TOKENS.IMAGE) {
      parsedContent.push(`<img src="${token.tokens.url} alt="${token.tokens.alt}">`)
      continue
    }
    
    // if checked checkbox
    if (token.type === TOKENS.CHECKBOX_CHECKED) {
      parsedContent.push(`<input type="checkbox" checked>${parseParaToken(token)}`)
    }
    
    // if paragraph
    if (token.type === TOKENS.PARAGRAPH) {
      parsedContent.push(
        `<p>${parseParaToken(token)}<p>`
      )
    }
  }
  // TODO: see why multiple br tags are added
  return parsedContent
    .map(item => item.replace("%s", ""))
    .join("\n")
}

module.exports = parser
