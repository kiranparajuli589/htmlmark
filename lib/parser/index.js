import { TOKENS } from "../util/tokens.js"
import { REGEX } from "../regex/index.js"
import { Table } from "../tokenizer/table.js"
import { Image } from "../tokenizer/image.js"
import { Comment } from "../tokenizer/comment.js"
import { HR } from "../tokenizer/hr.js"
import { CodeBlock } from "../tokenizer/codeblock.js"
import { Quote } from "../tokenizer/quote.js"
import { Heading } from "../tokenizer/heading.js"
import { List } from "../tokenizer/list.js"
import { Newline } from "../tokenizer/newline.js"


export class Parser {
	#lexers
	#cursor
	#parsedContent
	#currentLexer

	constructor(lexers) {
		this.#lexers = lexers
		this.#parsedContent = []
	}

	static escape(str) {
		return str.replaceAll(">", "&gt;")
			.replaceAll("<", "&lt;")
			.replaceAll("\"", "&quot;")
			.replaceAll("'", "&#39;")
	}

	static parseContent(token) {
		return token.value.replace(REGEX.PARAGRAPH.BOLD, "<strong>$1</strong>")
			.replace(REGEX.PARAGRAPH.ITALIC,"<em>$1</em>")
			.replace(REGEX.PARAGRAPH.CODE, (match, value) => {
				return `<code>${
					Parser.escape(value)
				}</code>`
			})
			.replace(REGEX.PARAGRAPH.STRIKE, "<s>$1</s>")
			.replace(REGEX.PARAGRAPH.UNDERLINE, "<u>$1</u>")
			.replace(REGEX.PARAGRAPH.LINK, "<a href='$2'>$1</a>")
	}

	#parseNewLine() {
		this.#parsedContent.push((Newline.parse()))
	}
	#parseHrLine() {
		this.#parsedContent.push(HR.parse())
	}
	#parseCodeBlock() {
		this.#parsedContent.push(CodeBlock.parse(this.#currentLexer))
	}
	#parseQuote() {
		this.#parsedContent.push(Quote.parse(this.#currentLexer))
	}
	#parseHeading() {
		this.#parsedContent.push(Heading.parse(this.#currentLexer))
	}
	#parseList() {
		this.#parsedContent.push(List.parse(this.#currentLexer))
	}
	#parseComment() {
		this.#parsedContent.push(Comment.parse(this.#currentLexer))
	}
	#parseImage() {
		this.#parsedContent.push(Image.parse(this.#currentLexer))
	}
	#parseTable() {
		this.#parsedContent.push(Table.parse(this.#currentLexer))
	}
	#parseParagraph() {
		this.#parsedContent.push(`<p>${Parser.parseContent(this.#currentLexer)}</p>`)
	}
	#parseCurrentLexer() {
		switch (this.#currentLexer.type) {
		case TOKENS.NEW_LINE:
			this.#parseNewLine()
			break
		case TOKENS.HR_LINE:
			this.#parseHrLine()
			break
		case TOKENS.CODE_BLOCK:
			this.#parseCodeBlock()
			break
		case TOKENS.QUOTE:
			this.#parseQuote()
			break
		case TOKENS.HEADING:
			this.#parseHeading()
			break
		case TOKENS.LIST:
			this.#parseList()
			break
		case TOKENS.COMMENT:
			this.#parseComment()
			break
		case TOKENS.IMAGE:
			this.#parseImage()
			break
		case TOKENS.TABLE:
			this.#parseTable()
			break
		case TOKENS.PARAGRAPH:
			this.#parseParagraph()
			break
		}
	}
	run() {
		for (this.#cursor=0; this.#cursor<this.#lexers.length; this.#cursor++) {
			this.#currentLexer = this.#lexers[this.#cursor]
			this.#parseCurrentLexer()
		}

		// remove any %s leftovers and return
		return this.#parsedContent
			.map(item => item.replaceAll("%s", ""))
			.join("")
	}
}
