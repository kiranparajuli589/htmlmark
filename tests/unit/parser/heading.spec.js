import { MDP } from "../../../lib"


describe("Heading Parsing", () => {
	it("should parse a fenced Heading", () => {
		const lines = [
			"# Heading #####",
			"# Heading###"
		]
		const html = MDP.h(lines)
		expect(html).toMatchSnapshot()
	})
})
