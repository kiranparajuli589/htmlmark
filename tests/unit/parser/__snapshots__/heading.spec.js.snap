// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Parsing should parse a fenced Heading 1`] = `
"<h1>Heading</h1>
<h1>Heading###</h1>
"
`;
